src/AutoRest.CSharp/Common/Output/Models/Types/SystemObjectType.cs

```diff
@@ -160,6 +161,37 @@ private bool GetReadOnly(PropertyInfo property)
             return property.GetSetMethod() == null;
         }
 
+        private bool GetRequired(PropertyInfo property)
+        {
+            var publicCtor = property.DeclaringType?.GetConstructors().Where(c => c.IsPublic).FirstOrDefault();
+            if (publicCtor == null)
+            {
+                //ReferenceType for inheritance do not have public constructor, and currently there are ResourceData, TrackedResourceData, WritableResourceData.
+                //We treat Id, Name, ResourceType, Location in ResourceData/TrackedResourceData as required.
+                if (!property.DeclaringType!.Name.Equals("WritableResourceData", StringComparison.Ordinal))
+                {
+                    if (property.Name == "Id" ||
+                        property.Name == "Name" ||
+                        property.Name == "ResourceType" ||
+                        property.Name == "Location")
+                        return true;
+                }
```

Comment:
The hard-code logic is actually for ResourceData/TrackedResourceData. Based on swagger definition, none of id, name, type is required and there is no public constructor which takes all required properties as parameters in the ReferenceTypes. Therefore, by default every property in ReferenceTypes is non-required. But since Azure.ResourceManager is GAed and in real life it seems reasonable to make id, name, type required in ResourceData/TrackedResourceData. So I added the hard-code logic here.

Reply:
I think you may have misunderstood the purpose of this piece of code, it's not for generating common types in ResourceManager (ResourceData and TrackedResourceData are all customized code). It's for building the SystemObjectType to replace the original type or use as base type in a model of some RP.  Here we build SystemObjectType from the common type models in ResourceManager, not from swagger. The challenge here is that we need to tell generator about whether the properties are required based on the ResourceData/TrackedResourceData classes themselves.